rust-symcrypt/src/ecdh.rs

```diff
@@ -0,0 +1,269 @@
+        }
+    }
+
+    /// EcDh::from_public_key_bytes()    function returns a EcDh struct that only has a public key attached.
+    pub fn from_public_key_bytes(
+        curve: CurveType,
+        public_key: &[u8],
+    ) -> Result<Self, SymCryptError> {
+        let mut instance = EcDh {
+            curve_type: curve,
+            expanded_curve: None,
+            key: None,
+        };
+
+        let num_format = get_num_format(instance.curve_type);
+
+        let ec_point_format = symcrypt_sys::_SYMCRYPT_ECPOINT_FORMAT_SYMCRYPT_ECPOINT_FORMAT_XY;
+
+        unsafe {
+            // SAFETY: FFI calls
+            let curve_ptr =
+                symcrypt_sys::SymCryptEcurveAllocate(convert_curve(instance.curve_type), 0);
+            if curve_ptr.is_null() {
+                return Err(SymCryptError::MemoryAllocationFailure);
+            }
+
+            instance.expanded_curve = Some(curve_ptr);
+
+            let key_ptr = symcrypt_sys::SymCryptEckeyAllocate(instance.expanded_curve.unwrap());
+            if key_ptr.is_null() {
+                return Err(SymCryptError::InvalidArgument);
+            }
+
+            match symcrypt_sys::SymCryptEckeySetValue(
+                std::ptr::null(),
+                0,
+                public_key.as_ptr(),
+                public_key.len() as symcrypt_sys::SIZE_T,
+                num_format,
+                ec_point_format,
+                symcrypt_sys::SYMCRYPT_FLAG_ECKEY_ECDH,
+                key_ptr,
+            ) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => {
+                    instance.key = Some(key_ptr);
+                    Ok(instance)
+                }
+                err => Err(err.into()),
+            }
+        }
+    }
+
+    pub fn get_public_key_bytes(&mut self) -> Result<Vec<u8>, SymCryptError> {
+        let num_format = get_num_format(self.curve_type);
+        let ec_point_format = symcrypt_sys::_SYMCRYPT_ECPOINT_FORMAT_SYMCRYPT_ECPOINT_FORMAT_XY;
+
+        unsafe {
+            // SAFETY: FFI calls
+            let pub_key_len = symcrypt_sys::SymCryptEckeySizeofPublicKey(
+                self.key.unwrap(),
+                symcrypt_sys::_SYMCRYPT_ECPOINT_FORMAT_SYMCRYPT_ECPOINT_FORMAT_XY,
+            );
+
+            let mut pub_key_bytes = vec![0u8; pub_key_len as usize];
+
+            match symcrypt_sys::SymCryptEckeyGetValue(
+                self.key.unwrap(),
+                std::ptr::null_mut(),
+                0 as symcrypt_sys::SIZE_T,
+                pub_key_bytes.as_mut_ptr(),
+                pub_key_len as symcrypt_sys::SIZE_T,
+                num_format,
+                ec_point_format,
+                0, // no flags allowed
+            ) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(pub_key_bytes),
+                err => Err(err.into()),
+            }
+        }
+    }
+
+    pub fn ecdh_secret_agreement(private: &EcDh, public: &EcDh) -> Result<Vec<u8>, SymCryptError> {
```

Comment:
Instead of returning `Vec<u8>` here, I'd consider returning something like `SecreteAgreement` which would be a type that wraps the array of bytes. It makes it easier to keep track of what the meaning of the bytes is elsewhere in the code.

Reply:
+1